spec/integration/avro_integration_spec.rb

```diff
@@ -0,0 +1,460 @@
+    let(:password) { "changeme" }
+
+    before(:all) do
+      run_integration_script("stop_schema_registry.sh")
+      sleep 2
+      run_integration_script("start_auth_schema_registry.sh")
+      sleep 5
+      wait_for_schema_registry("http://localhost:8081", username: "barney", password: "changeme")
+    end
+
+    after(:all) do
+      run_integration_script("stop_schema_registry.sh")
+      sleep 2
+    end
+
+    context "with valid credentials" do
+      let(:schema_subject) { "test-auth-#{Time.now.to_i}" }
+      let(:full_schema_url) { "#{schema_registry_url}/subjects/#{schema_subject}/versions/latest" }
+      let(:config) { super().merge({ 'schema_uri' => full_schema_url, 'username' => username, 'password' => password }) }
+
+      before do
+        client_options = { auth: { user: username, password: password } }
+        client = Manticore::Client.new(client_options)
+        response = client.post("#{schema_registry_url}/subjects/#{schema_subject}/versions",
+                               headers: { "Content-Type" => "application/vnd.schemaregistry.v1+json" },
+                               body: { schema: test_schema_json }.to_json
+        ).call
+        expect(response.code).to eq(200)
+        client.close
+      end
+
+      it "fetches schema with valid credentials" do
+        encoded_data = encode_avro_data(test_schema_json, test_event_data)
+        events = []
+        codec.decode(encoded_data) do |event|
+          events << event
+        end
+
+        expect(events.size).to eq(1)
+        expect(events.first.get("message")).to eq(test_event_data["message"])
+      end
+
+      it "encodes data with authentication" do
+        event = LogStash::Event.new(test_event_data)
+        encoded_data = nil
+
+        codec.on_event do |e, data|
+          encoded_data = data
+        end
+
+        codec.encode(event)
+        expect(encoded_data).not_to be_nil
+      end
+    end
+
+    context "with invalid credentials" do
+      let(:schema_subject) { "test-invalid-auth-#{Time.now.to_i}" }
+      let(:full_schema_url) { "#{schema_registry_url}/subjects/#{schema_subject}/versions/latest" }
+
+      before do
+        client_options = { auth: { user: username, password: password } }
+        client = Manticore::Client.new(client_options)
+        response = client.post("#{schema_registry_url}/subjects/#{schema_subject}/versions",
+                               headers: { "Content-Type" => "application/vnd.schemaregistry.v1+json" },
+                               body: { schema: test_schema_json }.to_json
+        ).call
+        expect(response.code).to eq(200)
+        client.close
+      end
+
+      it "fails with invalid credentials" do
+        expect {
+          invalid_config = { 'schema_uri' => full_schema_url, 'username' => 'invalid', 'password' => 'wrong' }
+          LogStash::Codecs::Avro.new(invalid_config).tap { |c| c.register }
+        }.to raise_error(/401|403|HTTP request failed/)
```

Comment:
I would expect 401 Unauthorized. Any reason to expect 403 or "HTTP request failed" ?

Reply:
Yeah I think being specific seems a better approach.
Kept only 401.